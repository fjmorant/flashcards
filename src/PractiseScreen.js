import React, {Component} from 'react'
import {Button, StyleSheet, Text, TouchableOpacity, View} from 'react-native'

import {graphql} from 'react-apollo'
import {flashcardsQuery, deleteFlashcardMutation} from './queries'

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 10,
  },
  closeContainer: {
    flex: 1,
    alignItems: 'flex-start',
  },
  flashCardsContainer: {
    flex: 15,
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'row',
  },
  flashcardHeaderContainer: {
    alignItems: 'center',
  },
  flashcardHeaderText: {
    fontSize: 30,
  },
  flashcardDefinitionContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    padding: 7,
  },
  flashcardDefinitionText: {
    fontSize: 25,
    textAlign: 'justify',
  },
  flashcardExampleContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    padding: 7,
  },
  flashcardExampleText: {
    fontSize: 25,
    textAlign: 'justify',
  },
  arrowText: {color: 'black', fontSize: 35},
  flashCardContainer: {flex: 1},
  loadingView: {flex: 1, justifyContent: 'center', alignItems: 'center'},
})

const FlashCard = ({
  name,
  meaning,
  mode,
  example,
}: {
  name: string,
  meaning: string,
  mode: string,
  example: string,
}) => (
  <View style={styles.flashCardContainer}>
    <View style={styles.flashcardHeaderContainer}>
      <Text style={styles.flashcardHeaderText}>{name}</Text>
    </View>
    {mode === 'DEFINITION' ? (
      <View style={styles.flashcardDefinitionContainer}>
        <Text style={styles.flashcardDefinitionText}>{meaning}</Text>
      </View>
    ) : null}
    {mode === 'EXAMPLE' ? (
      <View style={styles.flashcardExampleContainer}>
        <Text style={styles.flashcardExampleText}>{example}</Text>
      </View>
    ) : null}
  </View>
)

const Arrow = ({title, onPress}: {title: string, onPress: any}) => (
  <TouchableOpacity onPress={onPress}>
    <Text style={styles.arrowText}>{title}</Text>
  </TouchableOpacity>
)

export class PractiseScreen extends Component<{
  flashCardList: any,
  navigation: any,
  data: {
    loading: boolean,
    User: {
      flashcards: Array<any>,
    },
  },
}> {
  constructor() {
    super()

    this.state = {
      flashcardIndex: 0,
      mode: 'EXAMPLE',
    }
    this.goNextCard = this.goNextCard.bind(this)
    this.goPrevCard = this.goPrevCard.bind(this)
    this.switchMode = this.switchMode.bind(this)
  }

  componentWillMount() {
    if (!this.props.data.loading && !this.props.data.User.flashcards.length) {
      const {goBack} = this.props.navigation

      goBack()
      alert('No flash cards found!')
    }
  }

  static navigationOptions = ({navigation, screenProps}) => {
    return {
      header: null,
    }
  }

  goNextCard() {
    const numFlashcards = this.props.data.User.flashcards.length
    const oldIndex = this.state.flashcardIndex
    const flashcardIndex =
      oldIndex < numFlashcards - 1 ? oldIndex + 1 : numFlashcards - 1

    this.setState({flashcardIndex})
  }

  goPrevCard() {
    const flashcardIndex = (this.state.flashcardIndex || 1) - 1

    this.setState({flashcardIndex})
  }

  switchMode() {
    this.setState({
      mode: this.state.mode === 'DEFINITION' ? 'EXAMPLE' : 'DEFINITION',
    })
  }

  render() {
    const {goBack} = this.props.navigation
    const loading = this.props.data.loading

    if (loading) {
      return (
        <View style={styles.loadingView}>
          <Text>Loading</Text>
        </View>
      )
    }

    const flashcard = this.props.data.User.flashcards[this.state.flashcardIndex]
    const numFlashcards = this.props.data.User.flashcards.length

    return (
      <View style={styles.container}>
        <View style={styles.closeContainer}>
          <Button onPress={() => goBack()} title={'Close'} />
        </View>
        <View style={styles.flashCardsContainer}>
          <Arrow onPress={this.goPrevCard} title={'<'} />
          {flashcard ? (
            <FlashCard
                example={flashcard.example}
                meaning={flashcard.meaning}
                mode={this.state.mode}
                name={flashcard.name}
            />
          ) : null}
          <Arrow onPress={this.goNextCard} title={'>'} />
        </View>
        <Button
            onPress={this.switchMode}
            title={
            this.state.mode === 'DEFINITION' ? 'See Example' : 'See Definition'
          }
        />
      </View>
    )
  }
}

export default graphql(flashcardsQuery)(PractiseScreen)
