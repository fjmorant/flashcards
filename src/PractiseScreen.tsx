import * as React from 'react'
import {Button, StyleSheet, Text, TouchableOpacity, View} from 'react-native'

import {graphql} from 'react-apollo'
import {flashcardsQuery} from './queries'

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 10,
  },
  closeContainer: {
    flex: 1,
    alignItems: 'flex-start',
  },
  flashCardsContainer: {
    flex: 15,
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'row',
  },
  flashcardHeaderContainer: {
    alignItems: 'center',
  },
  flashcardHeaderText: {
    fontSize: 30,
  },
  flashcardDefinitionContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    padding: 7,
  },
  flashcardDefinitionText: {
    fontSize: 25,
    textAlign: 'justify',
  },
  flashcardExampleContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    padding: 7,
  },
  flashcardExampleText: {
    fontSize: 25,
    textAlign: 'justify',
  },
  arrowText: {color: 'black', fontSize: 35},
  flashCardContainer: {flex: 1},
  loadingView: {flex: 1, justifyContent: 'center', alignItems: 'center'},
})

// tslint:disable-next-line:variable-name
const FlashCard = ({
  name,
  meaning,
  mode,
  example,
}: {
  name: string
  meaning: string
  mode: string
  example: string
}) => (
  <View style={styles.flashCardContainer}>
    <View style={styles.flashcardHeaderContainer}>
      <Text style={styles.flashcardHeaderText}>{name}</Text>
    </View>
    {mode === 'DEFINITION' ? (
      <View style={styles.flashcardDefinitionContainer}>
        <Text style={styles.flashcardDefinitionText}>{meaning}</Text>
      </View>
    ) : null}
    {mode === 'EXAMPLE' ? (
      <View style={styles.flashcardExampleContainer}>
        <Text style={styles.flashcardExampleText}>{example}</Text>
      </View>
    ) : null}
  </View>
)

// tslint:disable-next-line:variable-name
const Arrow = ({title, onPress}: {title: string; onPress: any}) => (
  <TouchableOpacity onPress={onPress}>
    <Text style={styles.arrowText}>{title}</Text>
  </TouchableOpacity>
)

export interface IProps {
  flashCardList: any
  navigation: any
  data: {
    loading: boolean
    User: {
      flashcards: Array<any>
    }
  }
}

export interface IState {
  flashcardIndex: number
  mode: string
}

export class PractiseScreen extends React.Component<IProps, IState> {
  public static navigationOptions = () => {
    return {
      header: null,
    }
  }

  constructor(props: IProps) {
    super(props)

    this.state = {
      flashcardIndex: 0,
      mode: 'EXAMPLE',
    }
    this.goNextCard = this.goNextCard.bind(this)
    this.goPrevCard = this.goPrevCard.bind(this)
    this.switchMode = this.switchMode.bind(this)
  }

  public componentWillMount() {
    if (!this.props.data.loading && !this.props.data.User.flashcards.length) {
      const {goBack} = this.props.navigation

      goBack()
      alert('No flash cards found!')
    }
  }

  public goNextCard() {
    const numFlashcards = this.props.data.User.flashcards.length
    const oldIndex = this.state.flashcardIndex
    const flashcardIndex =
      oldIndex < numFlashcards - 1 ? oldIndex + 1 : numFlashcards - 1

    this.setState({flashcardIndex})
  }

  public goPrevCard() {
    const flashcardIndex = (this.state.flashcardIndex || 1) - 1

    this.setState({flashcardIndex})
  }

  public switchMode() {
    this.setState({
      mode: this.state.mode === 'DEFINITION' ? 'EXAMPLE' : 'DEFINITION',
    })
  }

  public render() {
    const {goBack} = this.props.navigation
    const loading = this.props.data.loading

    if (loading) {
      return (
        <View style={styles.loadingView}>
          <Text>Loading</Text>
        </View>
      )
    }

    const flashcard = this.props.data.User.flashcards[this.state.flashcardIndex]

    return (
      <View style={styles.container}>
        <View style={styles.closeContainer}>
          <Button onPress={() => goBack()} title={'Close'} />
        </View>
        <View style={styles.flashCardsContainer}>
          <Arrow onPress={this.goPrevCard} title={'<'} />
          {flashcard ? (
            <FlashCard
              example={flashcard.example}
              meaning={flashcard.meaning}
              mode={this.state.mode}
              name={flashcard.name}
            />
          ) : null}
          <Arrow onPress={this.goNextCard} title={'>'} />
        </View>
        <Button
          onPress={this.switchMode}
          title={
            this.state.mode === 'DEFINITION' ? 'See Example' : 'See Definition'
          }
        />
      </View>
    )
  }
}

export default graphql<
  {
    User: {
      id: string
      flashcards: Array<{
        id: string
        name: string
        example: string
        mastered: boolean
        meaning: string
      }>
    }
  },
  any
>(flashcardsQuery)(PractiseScreen)
