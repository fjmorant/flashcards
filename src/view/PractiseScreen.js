import React, {Component} from 'react';
import {
  Button,
  ScrollView,
  StyleSheet,
  Text,
  View,
} from 'react-native'

import immutable from 'immutable'
import {connect} from 'react-redux'

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 10,
  },
  closeContainer: {
    flex: 1,
    alignItems: 'flex-start',
  },
  flashCardsContainer: {
    flex: 15,
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'row',
  },
  flashcardHeaderContainer: {
    flex: 1,
    alignItems: 'center',
  },
  flashcardHeaderText: {
    fontSize: 30,
  },
  flashcardDefinitionContainer: {
    flex: 2,
    alignItems: 'center',
  },
  flashcardDefinitionText: {
    fontSize: 25,
  },
  flashcardExampleContainer: {
    flex: 2,
    alignItems: 'center',
  },
  flashcardExampleText: {
    fontSize: 25,
  }
})

const FlashCard = ({
  name,
  meaning,
  example
}) => (
  <View style={{flex: 1}}>
    <View style={styles.flashcardHeaderContainer}>
      <Text style={styles.flashcardHeaderText}>{name}</Text>
    </View>
    <View style={styles.flashcardDefinitionContainer}>
      <Text style={styles.flashcardDefinitionText}>
        {meaning}
      </Text>
    </View>
    <View style={styles.flashcardExampleContainer}>
      <Text style={styles.flashcardExampleText}>{example}</Text>
    </View>
  </View>
)

class PractiseScreen extends Component {

  constructor() {
    super()

    this.state = {
      flashcardIndex: 0,
    }
  }
  static navigationOptions = {
    header: {
      visible: false,
    }
  }

  render() {
    const {goBack} = this.props.navigation
    const flashcard = this.props.flashcards.get(this.state.flashcardIndex)
    const numFlashcards = this.props.flashcards.size

    return (
      <View style={styles.container}>
          <View style={styles.closeContainer}>
            <Button
              onPress={() => goBack()}
              title={'Close'}
            />
          </View>
          <View style={styles.flashCardsContainer}>
            <Button
              onPress={() => this.setState({
                flashcardIndex: (this.state.flashcardIndex || 1) - 1
              })}
              title={'Prev'}
            />
            {
              flashcard ? <FlashCard
                            name={flashcard.get('name')}
                            example={flashcard.get('example')}
                            meaning={flashcard.get('meaning')}
                          /> : null
            }
            <Button
              onPress={() => this.setState({
                flashcardIndex: this.state.flashcardIndex < numFlashcards - 1 ? this.state.flashcardIndex + 1 : numFlashcards - 1
              })}
              title={'Next'}
            />
          </View>
      </View>
    )
  }
}

export default connect(state => ({
    flashcards: state.flashcards.get('flashcards') || immutable.List(),
  }), null)(PractiseScreen);
