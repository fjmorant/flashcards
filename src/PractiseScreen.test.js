import 'react-native'
import React from 'react'
import {PractiseScreen} from './PractiseScreen'

import renderer from 'react-test-renderer'

jest.mock('react-dom/server', () => {}, {virtual: true})

describe('PractiseScreen', () => {
  test('renders correctly', () => {
    const tree = renderer
      .create(<PractiseScreen data={{loading: true}} navigation={{}} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
