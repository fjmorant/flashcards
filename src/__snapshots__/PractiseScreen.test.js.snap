// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PractiseScreen renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "padding": 10,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "flex": 1,
      }
    }
  >
    <View
      accessibilityComponentType="button"
      accessibilityLabel={undefined}
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={undefined}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Close
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 15,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "black",
            "fontSize": 35,
          }
        }
      >
        &lt;
      </Text>
    </View>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 30,
            }
          }
        >
          
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "padding": 7,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 25,
              "textAlign": "justify",
            }
          }
        >
          
        </Text>
      </View>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "black",
            "fontSize": 35,
          }
        }
      >
        &gt;
      </Text>
    </View>
  </View>
  <View
    accessibilityComponentType="button"
    accessibilityLabel={undefined}
    accessibilityTraits={
      Array [
        "button",
      ]
    }
    accessible={true}
    collapsable={undefined}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={undefined}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        See Definition
      </Text>
    </View>
  </View>
</View>
`;
